src/sage/combinat/abstract_tree.py

```diff
@@ -639,15 +638,127 @@ def action(x):
                 # subtrees, and should not be exploded again, but instead
                 # should be manipulated and removed from the stack.
                 stack.append(None)
-                for i in range(len(node)):
-                    subtree = node[-i - 1]
+                for subtree in reversed(node):
                     if not subtree.is_empty():
                         stack.append(subtree)
             else:
                 stack.pop()
                 node = stack.pop()
                 action(node)
 
+    def contour_traversal(self,first_action=None, middle_action=None, final_action=None, leaf_action=[]):
```

Comment:
```suggestion
    def contour_traversal(self, first_action=None, middle_action=None, final_action=None, leaf_action=None):
```
I don't see why `leaf_action` should not be `None` by default.

Reply:
I want that if you change only `first_action` and `final_action`, `leaf_action` will be the action of both of them but it's perhaps a bad idea.

Reply:
I don't understand, can you give an example?

Reply:
If you look at the depth code : i only change the `final_action` and it change accordingly the `leaf_action`. 
I would like it to work like that.

Reply:
I think again at this point and I think you are right, it's probably better that the default `leaf_action` is None. 
I made the change in the last commit.